src/main/java/org/isf/exam/rest/ExamController.java

```diff
@@ -0,0 +1,120 @@
+package org.isf.exam.rest;
+
+import io.swagger.annotations.Api;
+import io.swagger.annotations.Authorization;
+import org.isf.exa.manager.ExamBrowsingManager;
+import org.isf.exa.model.Exam;
+import org.isf.exam.dto.ExamDTO;
+import org.isf.exam.mapper.ExamMapper;
+import org.isf.exatype.manager.ExamTypeBrowserManager;
+import org.isf.exatype.model.ExamType;
+import org.isf.shared.exceptions.OHAPIException;
+import org.isf.utils.exception.OHServiceException;
+import org.isf.utils.exception.model.OHExceptionMessage;
+import org.isf.utils.exception.model.OHSeverityLevel;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.MediaType;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+import java.util.Optional;
+
+@RestController
+@Api(value = "/exams", produces = MediaType.APPLICATION_JSON_VALUE, authorizations = {@Authorization(value = "basicAuth")})
+public class ExamController {
+
+    private final Logger logger = LoggerFactory.getLogger(ExamController.class);
+
+    @Autowired
+    protected ExamBrowsingManager examManager;
+    @Autowired
+    protected ExamTypeBrowserManager examTypeBrowserManager;
+    @Autowired
+    private ExamMapper examMapper;
+
+    public ExamController(ExamBrowsingManager examManager, ExamMapper examMapper) {
+        this.examManager = examManager;
+        this.examMapper = examMapper;
+    }
+
+    @PostMapping(value = "/exams", produces = MediaType.APPLICATION_JSON_VALUE)
+    public ResponseEntity newExam(@RequestBody ExamDTO newExam) throws OHServiceException {
+        ExamType examType = examTypeBrowserManager.getExamType().stream().filter(et -> newExam.getExamtype().getCode().equals(et.getCode())).findFirst().orElse(null);
+
+        if (examType == null) {
+            throw new OHAPIException(new OHExceptionMessage(null, "Exam type not found!", OHSeverityLevel.ERROR));
+        }
+
+        Exam exam = examMapper.map2Model(newExam);
+        exam.setExamtype(examType);
+
+        boolean isCreated = examManager.newExam(exam);
+        if (!isCreated) {
+            throw new OHAPIException(new OHExceptionMessage(null, "Exam is not created!", OHSeverityLevel.ERROR));
+        }
+        return ResponseEntity.ok(exam.getCode());
+    }
+
+    @PutMapping(value = "/exams/{code:.+}", produces = MediaType.APPLICATION_JSON_VALUE)
+    public ResponseEntity updateExams(@PathVariable String code, @RequestBody ExamDTO updateExam) throws OHServiceException {
+
```

Comment:
@vicusotto Is it necessary to take the parameter exam _code_ here? Since lines 65-67 are used to check whether the code passed corresponds to the code in the _updateExam_ object and throw an exception if not. I think this data is redundant.

Reply:
I use to build PUT method with code as uri parameter. It seems a convention of Api Crud method. Found this [https://www.w3.org/Protocols/rfc2616/rfc2616.txt] HTTP standard docs, for the PUT method (9.6 PUT):

> The fundamental difference between the POST and PUT requests is
reflected in the different meaning of the Request-URI. The URI in a
POST request identifies the resource that will handle the enclosed
entity. That resource might be a data-accepting process, a gateway to
some other protocol, or a separate entity that accepts annotations.
In contrast, the URI in a PUT request identifies the entity enclosed
with the request -- the user agent knows what URI is intended and the
server MUST NOT attempt to apply the request to some other resource.
If the server desires that the request be applied to a different URI

Do you know it's a deprecated method usage?

Reply:
I think you're right. It's true that this way of doing things involves additional work to double-check the concordance of identifiers, but it complies with standards.